packages/grid/x-data-grid/src/hooks/features/pagination/useGridPagination.ts

```diff
@@ -44,17 +81,212 @@ export const useGridPagination = (
   apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
   props: Pick<
     DataGridProcessedProps,
-    | 'page'
-    | 'pageSize'
-    | 'onPageChange'
-    | 'onPageSizeChange'
+    | 'paginationModel'
+    | 'onPaginationModelChange'
     | 'autoPageSize'
     | 'rowCount'
     | 'initialState'
     | 'paginationMode'
+    | 'signature'
     | 'rowHeight'
   >,
 ) => {
-  useGridPageSize(apiRef, props);
-  useGridPage(apiRef, props);
+  const logger = useGridLogger(apiRef, 'useGridPagination');
+
+  const visibleTopLevelRowCount = useGridSelector(apiRef, gridVisibleTopLevelRowCountSelector);
+  const densityFactor = useGridSelector(apiRef, gridDensityFactorSelector);
+  const rowHeight = Math.floor(props.rowHeight * densityFactor);
+
+  apiRef.current.registerControlState({
+    stateId: 'pagination',
+    propModel: props.paginationModel,
+    propOnChange: props.onPaginationModelChange,
+    stateSelector: gridPaginationModelSelector,
+    changeEvent: 'paginationModelChange',
+  });
+
+  /**
+   * API METHODS
+   */
+  const setPage = React.useCallback<GridPaginationApi['setPage']>(
+    (page) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (page === currentModel.page) {
+        return;
+      }
+      logger.debug(`Setting page to ${page}`);
+      apiRef.current.setPaginationModel({ page, pageSize: currentModel.pageSize });
+    },
+    [apiRef, logger],
+  );
+
+  const setPageSize = React.useCallback<GridPaginationApi['setPageSize']>(
+    (pageSize) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (pageSize === currentModel.pageSize) {
+        return;
+      }
+
+      logger.debug(`Setting page size to ${pageSize}`);
+      apiRef.current.setPaginationModel({ pageSize, page: currentModel.page });
+    },
+    [apiRef, logger],
+  );
+
+  const setPaginationModel = React.useCallback<GridPaginationApi['setPaginationModel']>(
+    (paginationModel) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (paginationModel === currentModel) {
```

Comment:
Since you're comparing object, I doubt the equality will be `true`. Maybe better to test the values, but it woull expose to bug if we add other properties latter

```suggestion
      if (paginationModel.page === currentModel.page && paginationModel.pageSize === currentModel.pageSize) {
```

Reply:
Usually, I'm doing a simple equality check in the model updater and doing the smarter checks in each function.
See the filter plugin.

For the pagination, the state is simple enough so that the detailed check is doable.
But as you said, I think it's risky for the future.

I would prefer to keep exposing `setPage` / `setPageSize` / `setPaginationModel` and have each method check what is relevant.
Updating only one of the two is probably the most common scenario for users, so it should not be more complicated to do it than before